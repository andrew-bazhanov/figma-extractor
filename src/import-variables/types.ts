import type { FileNodesResponse, FullStyleMetadata } from 'figma-js';

export interface ImportVariablesAdapter {
  readonly data: unknown;
  get styleMetadata(): FullStyleMetadata[];
  get fileNodes(): FileNodesResponse;
}

export interface ImportVariablesValidatorData {
  validateData: () => Error | undefined;
}

export type DeepWriteable<T> = { -readonly [P in keyof T]: DeepWriteable<T[P]> };
