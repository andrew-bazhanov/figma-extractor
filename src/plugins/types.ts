import { Core, FigmaData } from '../core';

export interface Plugin {
  (core: Core, data: FigmaData): Promise<void> | void;
  pluginName: string;
}
