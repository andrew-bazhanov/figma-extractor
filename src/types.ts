type SVGCallback = (config?: import('svgo').Config) => import('svgo').Config;

export type IconConfig = {
  disabled: boolean;
  nodeIds: string[];
  iconName: (nameFromFigma: string) => string;
  skipIcon?: (name: string) => boolean;
  exportPath: string;
  generateSprite: boolean;
  generateTypes: boolean;
  localIcons?: boolean;
  optimizeSvg?: false | SVGCallback;
};

export type Config = {
  apiKey: string;
  fileId: string;
  styles: {
    exportPath: string;
    allowedThemes?: string[];
    defaultTheme?: string;
    colors?: {
      disabled: boolean;
      keyName?: (name?: string, useTheme?: boolean) => string;
      useTheme?: boolean;
    };
    gradients?: {
      disabled: boolean;
      keyName?: (name?: string) => string;
    };
    effects?: {
      disabled: boolean;
      keyName?: (name?: string, useTheme?: boolean) => string;
      useTheme?: boolean;
    };
    textStyles?: {
      disabled: boolean;
      keyName?: (nameFromFigma: string) => string;
      merge?: boolean;
    };
  };
  icons: IconConfig | IconConfig[];
  screens?: {
    [title: string]: number | string;
  };
};

export type OnlyArgs = 'colors' | 'effects' | 'textStyles' | 'gradients' | 'icons';
