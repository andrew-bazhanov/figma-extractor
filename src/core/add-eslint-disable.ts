const maxLineRuleComment = '/* eslint-disable max-lines */';

const namingConventionRuleComment = '/* eslint-disable @typescript-eslint/naming-convention */';

export const addEslintDisableAtTheTopOfText = (
  content: string,
  options: {
    disabledMaxLines: boolean;
    disabledTypescriptNamingConvention: boolean;
  },
) => {
  return `
    ${options.disabledMaxLines ? maxLineRuleComment : ''}
    ${options.disabledTypescriptNamingConvention ? namingConventionRuleComment : ''}
    ${content}
    `;
};
