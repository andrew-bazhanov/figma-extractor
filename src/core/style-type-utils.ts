import { StyleType } from 'figma-js';

// eslint-disable-next-line @typescript-eslint/naming-convention
type StyleTypePredicate = (item: { style_type: StyleType }) => boolean;

const getStyleTypePredicate = (styleType: StyleType): StyleTypePredicate => {
  // eslint-disable-next-line @typescript-eslint/naming-convention
  return ({ style_type }) => style_type === styleType;
};

export const styleTypeUtils = {
  isFill: getStyleTypePredicate('FILL'),
  isEffect: getStyleTypePredicate('EFFECT'),
  isText: getStyleTypePredicate('TEXT'),
};
