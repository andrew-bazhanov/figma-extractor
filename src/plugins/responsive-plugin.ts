import path from 'path';

import { getResponsive } from '../lib/get-responsive';
import { Plugin } from './types';

export const responsivePlugin: Plugin = (
  { config, writeFile, addEslintDisableRules, log },
  { variables },
) => {
  if (config?.styles?.responsive?.disabled) {
    log(
      '[info:responsive] >>> ',
      'Responsive plugin has been disabled so the plugin had not been launched',
    );

    return;
  }
  log('[info:responsive] >>> ', 'Responsive plugin starts working...');

  const { screens, ...result } = getResponsive(variables, config);

  writeFile(
    addEslintDisableRules(
      `
    const responsiveVariants = ${JSON.stringify(result)};

    module.exports = {responsiveVariants}`,
      ['disable-max-lines', 'disable-typescript-naming-convention'],
    ),
    path.join(config?.styles?.exportPath || '', 'responsive.js'),
  );
};

responsivePlugin.pluginName = 'responsive';
