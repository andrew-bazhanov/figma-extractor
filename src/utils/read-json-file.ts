import fs from 'fs';

export function readJsonFile<T = any>(filePath: string): T {
  try {
    const jsonData = fs.readFileSync(filePath, 'utf-8');

    return JSON.parse(jsonData);
  } catch (error) {
    console.error('Error reading the JSON file:', error);
    throw error; // rethrow the error for further handling if needed
  }
}
