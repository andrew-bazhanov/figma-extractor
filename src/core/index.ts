import { ClientInterface, FileNodesResponse, FullStyleMetadata } from 'figma-js';

import { Plugin } from '../plugins';
import { runFormattingFile } from './run-formatting-file';
import { styleTypeUtils } from './style-type-utils';
import { writeJsFile } from './write-js-file';

export interface FigmaData {
  figmaClient: ClientInterface;
  styleMetadata: readonly FullStyleMetadata[];
  fileNodes: FileNodesResponse;
}

export interface Core {
  config: Config;
  rootPath: string;
  plugins: Plugin[];
  styleTypeUtils: typeof styleTypeUtils;
  writeJsFile: typeof writeJsFile;
  runFormattingFile: typeof runFormattingFile;
}

export const createCore = (args: { config: Config; rootPath: string; plugins: Plugin[] }): Core => {
  return {
    ...args,
    styleTypeUtils,
    writeJsFile,
    runFormattingFile: runFormattingFile,
  };
};
