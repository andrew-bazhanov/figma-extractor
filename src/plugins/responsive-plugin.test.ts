/* eslint-disable max-lines */
import path from 'path';

import { createCore } from '../core';
import { ThemeVariablesConfig } from '../types';
import { readJsonFile } from '../utils/read-json-file';
import { responsivePlugin } from './responsive-plugin';

describe('responsivePlugin', () => {
  const rootPath = process.cwd();
  const jsonVariablesPath = './variables.json';
  const variables = readJsonFile<ThemeVariablesConfig[]>(path.join(rootPath, jsonVariablesPath));

  it('should write responsive js file and run formatting tool', () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          responsive: { collectionNames: ['responsive', 'responsive_extra'] },
        },
      },
      plugins: [],
      rootPath: '/root-path',
      log: jest.fn(),
    });

    core.writeFile = jest.fn();

    responsivePlugin(core, { variables } as any);

    expect(core.writeFile).toHaveBeenCalled();
    expect((core.writeFile as jest.Mock).mock.calls[0][1]).toBe('/export-path/responsive.ts');
  });

  it('should write screens js file and run formatting tool', () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          responsive: { collectionNames: ['responsive', 'responsive_extra'] },
        },
      },
      plugins: [],
      rootPath: '/root-path',
      log: jest.fn(),
    });

    core.writeFile = jest.fn();

    responsivePlugin(core, { variables } as any);
    expect(core.writeFile).toHaveBeenCalled();
    expect((core.writeFile as jest.Mock).mock.calls[1][1]).toBe('/export-path/screens.ts');
  });
});
