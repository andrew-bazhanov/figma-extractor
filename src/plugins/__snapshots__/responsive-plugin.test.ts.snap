// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`responsivePlugin should write responsive js file and run formatting tool: /export-path/responsive.ts 1`] = `
"
    /* eslint-disable max-lines */
    /* eslint-disable @typescript-eslint/naming-convention */
    export const responsiveVariants = {".box-padding-px-200":{"padding-left":"8px","padding-right":"8px","@media (min-width: 680px)":{"padding-left":"8px","padding-right":"8px"},"@media (min-width: 1024px)":{"padding-left":"8px","padding-right":"8px"},"@media (min-width: 1440px)":{"padding-left":"8px","padding-right":"8px"}},".box-padding-py-200":{"padding-top":"8px","padding-bottom":"8px","@media (min-width: 680px)":{"padding-top":"8px","padding-bottom":"8px"},"@media (min-width: 1024px)":{"padding-top":"8px","padding-bottom":"8px"},"@media (min-width: 1440px)":{"padding-top":"8px","padding-bottom":"8px"}},".box-padding-px-600":{"padding-left":"24px","padding-right":"24px","@media (min-width: 680px)":{"padding-left":"24px","padding-right":"24px"},"@media (min-width: 1024px)":{"padding-left":"24px","padding-right":"24px"},"@media (min-width: 1440px)":{"padding-left":"24px","padding-right":"24px"}},".box-padding-py-600":{"padding-top":"24px","padding-bottom":"24px","@media (min-width: 680px)":{"padding-top":"24px","padding-bottom":"24px"},"@media (min-width: 1024px)":{"padding-top":"24px","padding-bottom":"24px"},"@media (min-width: 1440px)":{"padding-top":"24px","padding-bottom":"24px"}},".box-gap-300":{"gap":"24px","@media (min-width: 680px)":{"gap":"32px"},"@media (min-width: 1024px)":{"gap":"40px"},"@media (min-width: 1440px)":{"gap":"48px"}},".box-gap-200":{"gap":"24px","@media (min-width: 680px)":{"gap":"24px"},"@media (min-width: 1024px)":{"gap":"32px"},"@media (min-width: 1440px)":{"gap":"40px"}},".box-gap-100":{"gap":"8px","@media (min-width: 680px)":{"gap":"8px"},"@media (min-width: 1024px)":{"gap":"16px"},"@media (min-width: 1440px)":{"gap":"16px"}},".box-gap-100-negative":{"gap":"-8px","@media (min-width: 680px)":{"gap":"-8px"},"@media (min-width: 1024px)":{"gap":"-8px"},"@media (min-width: 1440px)":{"gap":"-8px"}},".box-gap-200-negative":{"gap":"-16px","@media (min-width: 680px)":{"gap":"-16px"},"@media (min-width: 1024px)":{"gap":"-24px"},"@media (min-width: 1440px)":{"gap":"-24px"}},".box-max-width-100":{"max-width":"400px","@media (min-width: 680px)":{"max-width":"400px"},"@media (min-width: 1024px)":{"max-width":"400px"},"@media (min-width: 1440px)":{"max-width":"400px"}},".box-min-width-100":{"min-width":"240px","@media (min-width: 680px)":{"min-width":"240px"},"@media (min-width: 1024px)":{"min-width":"240px"},"@media (min-width: 1440px)":{"min-width":"240px"}},".box-min-width-200":{"min-width":"320px","@media (min-width: 680px)":{"min-width":"320px"},"@media (min-width: 1024px)":{"min-width":"320px"},"@media (min-width: 1440px)":{"min-width":"320px"}},".box-min-width-300":{"min-width":"400px","@media (min-width: 680px)":{"min-width":"400px"},"@media (min-width: 1024px)":{"min-width":"400px"},"@media (min-width: 1440px)":{"min-width":"400px"}},".box-max-width-200":{"max-width":"600px","@media (min-width: 680px)":{"max-width":"600px"},"@media (min-width: 1024px)":{"max-width":"600px"},"@media (min-width: 1440px)":{"max-width":"600px"}},".box-max-width-300":{"max-width":"800px","@media (min-width: 680px)":{"max-width":"800px"},"@media (min-width: 1024px)":{"max-width":"800px"},"@media (min-width: 1440px)":{"max-width":"800px"}},".container-padding-px-600":{"padding-left":"16px","padding-right":"16px","@media (min-width: 680px)":{"padding-left":"24px","padding-right":"24px"},"@media (min-width: 1024px)":{"padding-left":"40px","padding-right":"40px"},"@media (min-width: 1440px)":{"padding-left":"80px","padding-right":"80px"}},".container-padding-py-600":{"padding-top":"16px","padding-bottom":"16px","@media (min-width: 680px)":{"padding-top":"24px","padding-bottom":"24px"},"@media (min-width: 1024px)":{"padding-top":"40px","padding-bottom":"40px"},"@media (min-width: 1440px)":{"padding-top":"80px","padding-bottom":"80px"}},".container-padding-px-200":{"padding-left":"16px","padding-right":"16px","@media (min-width: 680px)":{"padding-left":"24px","padding-right":"24px"},"@media (min-width: 1024px)":{"padding-left":"24px","padding-right":"24px"},"@media (min-width: 1440px)":{"padding-left":"24px","padding-right":"24px"}},".container-padding-py-200":{"padding-top":"16px","padding-bottom":"16px","@media (min-width: 680px)":{"padding-top":"24px","padding-bottom":"24px"},"@media (min-width: 1024px)":{"padding-top":"24px","padding-bottom":"24px"},"@media (min-width: 1440px)":{"padding-top":"24px","padding-bottom":"24px"}},".container-gap-100":{"gap":"16px","@media (min-width: 680px)":{"gap":"24px"},"@media (min-width: 1024px)":{"gap":"40px"},"@media (min-width: 1440px)":{"gap":"80px"}},".container-gap-100-negative":{"gap":"-40px","@media (min-width: 680px)":{"gap":"-40px"},"@media (min-width: 1024px)":{"gap":"-40px"},"@media (min-width: 1440px)":{"gap":"-40px"}},".container-max-width-100":{"max-width":"679px","@media (min-width: 680px)":{"max-width":"1023px"},"@media (min-width: 1024px)":{"max-width":"1439px"},"@media (min-width: 1440px)":{"max-width":"1920px"}},".container-min-width-100":{"min-width":"320px","@media (min-width: 680px)":{"min-width":"680px"},"@media (min-width: 1024px)":{"min-width":"1024px"},"@media (min-width: 1440px)":{"min-width":"1440px"}},".section-padding-px-100":{"padding-left":"24px","padding-right":"24px","@media (min-width: 680px)":{"padding-left":"32px","padding-right":"32px"},"@media (min-width: 1024px)":{"padding-left":"64px","padding-right":"64px"},"@media (min-width: 1440px)":{"padding-left":"80px","padding-right":"80px"}},".section-padding-py-100":{"padding-top":"24px","padding-bottom":"24px","@media (min-width: 680px)":{"padding-top":"32px","padding-bottom":"32px"},"@media (min-width: 1024px)":{"padding-top":"0px","padding-bottom":"0px"},"@media (min-width: 1440px)":{"padding-top":"0px","padding-bottom":"0px"}},".section-padding-py-700":{"padding-top":"80px","padding-bottom":"80px","@media (min-width: 680px)":{"padding-top":"80px","padding-bottom":"80px"},"@media (min-width: 1024px)":{"padding-top":"120px","padding-bottom":"120px"},"@media (min-width: 1440px)":{"padding-top":"120px","padding-bottom":"120px"}},".section-gap-100":{"gap":"24px","@media (min-width: 680px)":{"gap":"32px"},"@media (min-width: 1024px)":{"gap":"0px"},"@media (min-width: 1440px)":{"gap":"0px"}},".section-min-height-100vh":{"0":"0","1":"p","2":"x","@media (min-width: 680px)":"0px","@media (min-width: 1024px)":"0px","@media (min-width: 1440px)":"0px"},".screens-padding-px-100":{"padding-left":"24px","padding-right":"24px","@media (min-width: 680px)":{"padding-left":"24px","padding-right":"24px"},"@media (min-width: 1024px)":{"padding-left":"80px","padding-right":"80px"},"@media (min-width: 1440px)":{"padding-left":"120px","padding-right":"120px"}},".screens-padding-py-100":{"padding-top":"24px","padding-bottom":"24px","@media (min-width: 680px)":{"padding-top":"24px","padding-bottom":"24px"},"@media (min-width: 1024px)":{"padding-top":"80px","padding-bottom":"80px"},"@media (min-width: 1440px)":{"padding-top":"120px","padding-bottom":"120px"}}};
    "
`;

exports[`responsivePlugin should write screens js file and run formatting tool: /export-path/screens.ts 1`] = `
"
    /* eslint-disable max-lines */
    /* eslint-disable @typescript-eslint/naming-convention */
    export const screens = {"bs":"320px","sm":"680px","md":"1024px","lg":"1440px","xl":"1024px","2xl":"1440px"};
    "
`;
